### PR TITLE
Solving optimizer issues & typos

### PR DESCRIPTION
**Pull request by cetagostini**
## Description

We got a few concerns regarding the optimizer in the [forum](https://discourse.pymc.io/t/mmm-allocate-budget-to-maximize-response-generating-weird-decimal-places-results/14828/10?u=cluhmann) and the [discussions](https://github.com/pymc-labs/pymc-marketing/discussions/843) regarding the response the optimizer. Deeper checks reveal an scale problem once we transform back the response from the optimizer.

On the other hand we got a [typo](https://discourse.pymc.io/t/pymc-marketing-question-on-total-budget-parameter-for-budget-allocation-model/14259/2) in the function discussed here.

Finally, we where adding implicit noise to the final optimization. We added the noise level parameter to let the user to control if they want or not to add noise to their estimations.

## Related Issue

- [x] Closes #843 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [ ] New feature / enhancement
- [x] Bug fix
- [x] Documentation
- [x] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--933.org.readthedocs.build/en/933/

### CODE REVIEW

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (juanitorduz) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:2164]: Is there a way to add a test to ensure this change is working as expected?

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cluhmann) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:2112]: Maybe refer to these as "periods" because that's what the description of `time_granularity` refers to (assuming that `horizon` is measured in units of `time_granularity`)?  So:

```
horizon : int
    The number of time periods over which the budget is to be allocated.
```

Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:2112]: Correct, I'll replace ✌🏻

Comment (cluhmann) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:2112]: Also, I am not sure what standard procedure for the project is given that we're at v0.8.0 at this point, but because this is a breaking change, I would typically recommend throwing a warning if the old argument name is used (I mean, usually you would want through a whole deprecation cycle with  future warnings, etc. but probably not necessary for versions < 1.0).

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:2164]: Totally agree, just added three test to verify if the input budget and estimated budget are the same. Same thing with bounds, and periods.

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cluhmann) [inline pymc_marketing/mmm/budget_optimizer.py:54]: Number of days or the number of periods?  Also, why `periods` and not `num_periods`?  The latter seems less ambiguous.

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cluhmann) [inline pymc_marketing/mmm/budget_optimizer.py:71]: This seems to suggest that `periods` is in units of periods, not days.

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cluhmann) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:2111]: Same comment as above. Why `periods` instead of `num_periods`?

**Review by cluhmann — COMMENTED** [review]
Focused on the code rather than the notebooks (which seem like they were simply re-executed). The test also seems ok, but I am not super confident in my evaluation of it.  Left a couple of comments about argument names and docstrings.

**Comment by juanitorduz** [comment]
We have the following test failing because the change of parameter name for the budget optimizer 

```python
FAILED tests/mmm/test_budget_optimizer.py::test_allocate_budget[default_minimizer_kwargs]
FAILED tests/mmm/test_budget_optimizer.py::test_allocate_budget[custom_minimizer_kwargs]
FAILED tests/mmm/test_budget_optimizer.py::test_allocate_budget_zero_total[0-budget_bounds0-parameters0-expected_optimal0-2.38e-10]
FAILED tests/mmm/test_budget_optimizer.py::test_allocate_budget_custom_minimize_args
FAILED tests/mmm/test_budget_optimizer.py::test_allocate_budget_infeasible_constraints[100-budget_bounds0-parameters0-custom_constraints0]
```

We need to update the tests so that the optimizer uses  `periods` instead of `num_days`.

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cetagostini) [reply pymc_marketing/mmm/budget_optimizer.py:54]: Okay trying to be more explicit ✌🏻

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py]: Made the changes!

**Comment by cetagostini** [comment]
@juanitorduz the test should be done now!

@cluhmann the new description should be more understandable?

**Review by juanitorduz — APPROVED** [review]
Thank you @cetagostini !

**Comment by cetagostini** [comment]
@juanitorduz Thanks to you, will wait for Chris opinion before merge!

File: pymc_marketing/mmm/budget_optimizer.py
Comment (cluhmann) [inline pymc_marketing/mmm/budget_optimizer.py:54]: In general, I think this doc string (not just this line) could use with a bit more elaboration. For example, the number of time units...what? But given that this is largely a bug fix, I say merge it and we can improve docstrings later.

**Review by cluhmann — APPROVED** [review]

**Pull request merged**
